In [3]:
import plotly.io as pio
pio.renderers.default = "vscode+jupyterlab+notebook_connected"

# **Project 1: Bank Return on Equity of Countries**

## **Project Overview**
This project focuses on analyzing the "Bank return on equity (%, after tax)" metric for the year 2010, obtained from the World Bank's Global Financial Development database. The analysis aims to compare the after-tax return on equity for banks across various countries, providing insights into the financial performance of banking sectors worldwide.

Citation: [The World Bank - Data Bank | Global Financial Development](https://databank.worldbank.org/source/global-financial-development#).

## **1. Using Pandas**

**Step 1-1: Read in the Data**

Load the dataset using pandas, specifying the ISO-8859-1 encoding to handle any special characters in the file.

In [4]:
import pandas as pd

data = pd.read_csv('26259f8e-42d8-481c-9782-49cb6eed8ec1_Data.csv')

**Step 1-2: Data Cleaning**

Replace ".." with NaN to facilitate numeric processing, and display the first few rows to verify the changes.

In [5]:
import numpy as np
data.replace("..", np.nan, inplace=True)
data.head()

Country Name Country Code                                   Series Name  \
0  Afghanistan          AFG  Domestic credit to private sector (% of GDP)   
1  Afghanistan          AFG              Bank capital to total assets (%)   
2  Afghanistan          AFG  Bank non-performing loans to gross loans (%)   
3  Afghanistan          AFG          Bank return on equity (%, after tax)   
4  Afghanistan          AFG          International debt issues to GDP (%)   

  Series Code 2012 [YR2012] 2013 [YR2013] 2014 [YR2014] 2015 [YR2015]  \
0  GFDD.DI.14        4.4721       4.32169       3.94106       3.98149   
1  GFDD.SI.03           NaN           NaN           NaN           NaN   
2  GFDD.SI.02           NaN           NaN           NaN           NaN   
3  GFDD.EI.06      11.36923      8.266483      6.641262    -0.1358924   
4  GFDD.DM.07           NaN           NaN           NaN           NaN   

  2016 [YR2016] 2017 [YR2017] 2018 [YR2018] 2019 [YR2019] 2020 [YR2020]  \
0       3.88832       3.84188       3.51232       3.22996       3.07194   
1           NaN           NaN           NaN           NaN           NaN   
2           NaN           NaN           NaN           NaN           NaN   
3      14.81236      5.103842      6.225445      8.408169      3.150429   
4           NaN           NaN           NaN           NaN           NaN   

  2021 [YR2021]  
0           NaN  
1           NaN  
2           NaN  
3      11.39836  
4           NaN

**Step 1-3: Select the Data**

Use the following code to select and display the columns "Country Name," "Country Code," "Series Name," and "2020 [YR2020]."

In [6]:
data[['Country Name', 'Country Code', 'Series Name', '2020 [YR2020]']].head()

Country Name Country Code                                   Series Name  \
0  Afghanistan          AFG  Domestic credit to private sector (% of GDP)   
1  Afghanistan          AFG              Bank capital to total assets (%)   
2  Afghanistan          AFG  Bank non-performing loans to gross loans (%)   
3  Afghanistan          AFG          Bank return on equity (%, after tax)   
4  Afghanistan          AFG          International debt issues to GDP (%)   

  2020 [YR2020]  
0       3.07194  
1           NaN  
2           NaN  
3      3.150429  
4           NaN

Filter the dataset to extract entries where the "Series Name" is "Bank return on equity (%, after tax)," and display the relevant columns

In [7]:
bank_roe = data[data['Series Name'] == 'Bank return on equity (%, after tax)']

bank_roe[['Country Name', 'Country Code', 'Series Name', '2020 [YR2020]']]

Country Name Country Code  \
3               Afghanistan          AFG   
12                  Albania          ALB   
21                  Algeria          DZA   
30                  Andorra          AND   
39                   Angola          AGO   
...                     ...          ...   
1983  Virgin Islands (U.S.)          VIR   
1992     West Bank and Gaza          PSE   
2001            Yemen, Rep.          YEM   
2010                 Zambia          ZMB   
2019               Zimbabwe          ZWE   

                               Series Name 2020 [YR2020]  
3     Bank return on equity (%, after tax)      3.150429  
12    Bank return on equity (%, after tax)       9.41525  
21    Bank return on equity (%, after tax)      8.307635  
30    Bank return on equity (%, after tax)      5.728261  
39    Bank return on equity (%, after tax)     -16.76851  
...                                    ...           ...  
1983  Bank return on equity (%, after tax)           NaN  
1992  Bank return on equity (%, after tax)           NaN  
2001  Bank return on equity (%, after tax)           NaN  
2010  Bank return on equity (%, after tax)      7.767215  
2019  Bank return on equity (%, after tax)      10.22939  

[225 rows x 4 columns]

**Step 1-4: Computing the Mean, Median, and Mode**

Remove any rows containing "NaN" values from the "2020 [YR2020]" column to ensure accurate calculations.

In [8]:
bank_roe_sorted = bank_roe.dropna(subset=['2020 [YR2020]']).sort_values(by='Country Code')
bank_roe_sorted[['Country Name', 'Country Code', 'Series Name', '2020 [YR2020]']]

Country Name Country Code                           Series Name  \
3              Afghanistan          AFG  Bank return on equity (%, after tax)   
39                  Angola          AGO  Bank return on equity (%, after tax)   
12                 Albania          ALB  Bank return on equity (%, after tax)   
30                 Andorra          AND  Bank return on equity (%, after tax)   
1911  United Arab Emirates          ARE  Bank return on equity (%, after tax)   
...                    ...          ...                                   ...   
1965         Venezuela, RB          VEN  Bank return on equity (%, after tax)   
1974               Vietnam          VNM  Bank return on equity (%, after tax)   
1650          South Africa          ZAF  Bank return on equity (%, after tax)   
2010                Zambia          ZMB  Bank return on equity (%, after tax)   
2019              Zimbabwe          ZWE  Bank return on equity (%, after tax)   

     2020 [YR2020]  
3         3.150429  
39       -16.76851  
12         9.41525  
30        5.728261  
1911      5.195372  
...            ...  
1965      2.921044  
1974      15.31698  
1650      5.338054  
2010      7.767215  
2019      10.22939  

[152 rows x 4 columns]

The following code calculates the mean, median, and mode for the "Bank return on equity (%, after tax)" indicator for the year 2020.

In [9]:
mean_value = bank_roe_sorted['2020 [YR2020]'].astype(float).mean()
median_value = bank_roe_sorted['2020 [YR2020]'].astype(float).median()

mode_value_series = bank_roe_sorted['2020 [YR2020]'].astype(float).mode()
mode_value = mode_value_series.iloc[0] if not mode_value_series.empty else float('nan')

print(f"Mean: {mean_value}")
print(f"Median: {median_value}")
print(f"Mode: {mode_value}")


Mean: 8.529907913157894
Median: 7.462875
Mode: -16.76851


## **2. Using Pure Python**

**Step 2-1: Load and Preprocess Data**

**Library Imports**: The code imports the `csv` library to handle reading from the CSV file.

- **Initialize Storage for Values**: An empty list, `bank_roe_values`, is set up to store valid numerical values for bank return on equity (ROE) from the data.

- **Open and Read the CSV File**: The specified file is opened, and `csv.DictReader` reads each row as a dictionary.

- **Data Sorting and Filtering**:
  - Rows are sorted by "Country Code" for consistent output.
  - Only rows with the series name "Bank return on equity (%, after tax)" are selected.
  - Values that are either missing ("..") or empty are ignored. Valid numeric entries are converted to `float` and added to `bank_roe_values`.

In [10]:
import csv

file_path = '26259f8e-42d8-481c-9782-49cb6eed8ec1_Data.csv'

bank_roe_values = []
with open(file_path, mode='r', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    for row in sorted(reader, key=lambda x: x['Country Code']):
        if row['Series Name'] == 'Bank return on equity (%, after tax)':
            value = row['2020 [YR2020]']
            if value != ".." and value:
                try:
                    bank_roe_values.append(float(value))
                except ValueError:
                    continue


**Step 2-2: Computing the Mean, Median, and Mode**

The following code calculates the **mean**, **median**, and **mode** of the `bank_roe_values` list, which contains the Bank Return on Equity data. Each calculation is done manually without using external statistical libraries.

- **Calculate Mean**: The mean is computed as the sum of all values divided by the total count of values.

- **Calculate Median**:
  - The list is sorted in ascending order.
  - If the count of values (`n`) is even, the median is the average of the two middle values.
  - If `n` is odd, the median is the middle value.

- **Calculate Mode**:
  - A dictionary, `value_counts`, tracks the frequency of each value in the sorted list.
  - The mode is determined as the value(s) with the highest frequency. If multiple values have the same highest frequency, the first mode is selected.

In [11]:
# Calculate Mean
mean_value = sum(bank_roe_values) / len(bank_roe_values)

# Calculate Median
sorted_values = sorted(bank_roe_values)
n = len(sorted_values)
if n % 2 == 0:
    median_value = (sorted_values[n // 2 - 1] + sorted_values[n // 2]) / 2
else:
    median_value = sorted_values[n // 2]

# Calculate Mode
value_counts = {}
for value in sorted_values:
    if value in value_counts:
        value_counts[value] += 1
    else:
        value_counts[value] = 1
        
# Find the value(s) with the highest frequency
max_count = max(value_counts.values())
modes = [key for key, count in value_counts.items() if count == max_count]
mode_value = modes[0] if modes else None  # Choose the first mode if multiple

#Print
print(f"Mean: {mean_value}")
print(f"Median: {median_value}")
print(f"Mode: {mode_value}")


Mean: 8.529907913157894
Median: 7.462875
Mode: -16.76851


## **3. Data Visualization**

**Statistics Calculation for Bank Return on Equity**: The following code calculates the mean, median, and mode for the "Bank return on equity (%, after tax)" indicator for the year 2020.

- **Filter data for the specific metric and year, then sort by value for display**:
  - Filter the data for 'Bank return on equity (%, after tax)'

- **Determine maximum bar length**:
  - This keeps it compact for narrow screens 

- **Find the maximum value in the dataset to scale bars proportionally**:
  - Calculate the maximum absolute value for scaling the bars

- **Generate the textual bar chart**:
  - Scale the bar length relative to max_value
  - Format and print the country name and bar
  - Use '■' for positive and '□' for negative values

In [12]:
import pandas as pd
import numpy as np

file_path = '26259f8e-42d8-481c-9782-49cb6eed8ec1_Data.csv'
data = pd.read_csv(file_path)
data.replace("..", np.nan, inplace=True)

filtered_data = data[data['Series Name'] == 'Bank return on equity (%, after tax)']
filtered_data = filtered_data.dropna(subset=['2020 [YR2020]']).sort_values(by='2020 [YR2020]', ascending=True)
filtered_data['2020 [YR2020]'] = filtered_data['2020 [YR2020]'].astype(float)

max_bar_length = 10

max_value = abs(filtered_data['2020 [YR2020]'].min()) if filtered_data['2020 [YR2020]'].min() < 0 else filtered_data['2020 [YR2020]'].max()

print("Bank Return on Equity (%, 2020) by Country\n")
for index, row in filtered_data.iterrows():
    country = row['Country Name']
    value = row['2020 [YR2020]']
    
    bar_length = int((abs(value) / max_value) * max_bar_length)
    
    if value >= 0:
        bar = '■' * bar_length
        print(f"{country:30} | {bar} {value:.2f}")
    else:
        bar = '□' * bar_length
        print(f"{country:30} | {bar} {value:.2f}")


Bank Return on Equity (%, 2020) by Country

Poland                         |  -0.47
Italy                          |  -1.14
Cyprus                         | □ -1.82
Liberia                        | □□□□□□□□□ -15.32
Angola                         | □□□□□□□□□□ -16.77
Portugal                       | □ -2.03
Spain                          | □ -2.29
Bhutan                         | □ -2.83
Greece                         | □ -2.90
Trinidad and Tobago            | □□□ -5.05
Lebanon                        | □□□ -5.10
Congo, Dem. Rep.               | □□□ -5.36
Germany                        | □□□ -6.13
Ireland                        |  0.64
Austria                        |  1.66
Zimbabwe                       | ■■■■■■ 10.23
China                          | ■■■■■■ 10.25
Tunisia                        | ■■■■■■ 10.28
Nepal                          | ■■■■■■ 10.45
Norway                         | ■■■■■■ 10.46
Qatar                          | ■■■■■■ 10.67
Tajikistan                     | ■■■■■■ 10.7

## **4. Conclusion from the Data**

**Definition of the "Bank return on equity (%, after tax)"**

- **Commercial banks’ after-tax net income to yearly averaged equity**
  - Commercial banks’ after-tax net income to yearly averaged equity. Raw data are from Bankscope and Orbis. The formula applied to Bankscope is data2115[t] / ((data2055[t] + data2055[t-1])/2) and a comparable approach is applied to Orbis. Numerator and denominator are first aggregated on the country level before division. Note that banks used in the calculation might differ between indicators. Calculated from underlying bank-by-bank unconsolidated data from Bankscope and Orbis. The result is not reported if a country-year has less than 3 bank-level observations.

**Conclusion**
  - The analysis of the Bank Return on Equity (ROE) for the year 2020 reveals a diverse landscape across various countries. The global mean ROE stands at 8.53%, with a median of 7.46%, indicating that while many countries performed relatively well, several outliers demonstrate significantly lower returns, particularly in Europe and some developing nations.
  - Notably, several countries reported negative ROE, highlighting challenges faced by their banking sectors. For example, Liberia (-15.32%), Angola (-16.77%), and Italy (-1.14%) are among those struggling to generate profit relative to their equity base. Such figures suggest underlying issues such as economic instability, high operational costs, or poor asset quality that adversely affect bank profitability.
  - In contrast, numerous countries exhibit robust performance, with the highest ROE reported by Syrian Arab Republic (77.11%). Other countries, including Guinea (30.25%) and Burundi (31.51%), also show exceptionally high returns, indicating effective management and strong market conditions.
  - Overall, the distribution of ROE across countries highlights significant disparities in banking performance, influenced by local economic conditions, regulatory environments, and banking practices. This data underscores the importance of assessing country-specific factors when analyzing bank profitability and the overall health of financial systems globally.